feat: add tracestate header when starting lambda transaction

astorm opened the pull request:
### Checklist

Part of https://github.com/elastic/apm-agent-nodejs/issues/2156

This PR ensures that if the Lambda function is responding to an API Gateway request (or other request that includes a `tracestate` header) that its transaction will start using that tracestate.  This is needed because our existing lambda instrumentation was written before `tracestate` was a thing.

<!-- Potential tasks related to a new PR. Remove tasks that are not relevant -->

- [x] Implement code
- [x] Add tests

apmmachine commented:
## :grey_exclamation: Build Aborted
  
  > There is a new build on-going so the previous on-going builds have been aborted.
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-2185/detail/PR-2185/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-2185/detail/PR-2185/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-2185/detail/PR-2185/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-2185/detail/PR-2185/1//artifacts)   [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/apm-ci/transactions/view?rangeFrom=2021-08-09T20:31:11.918Z&rangeTo=2021-08-09T20:51:11.918Z&transactionName=apm-agent-nodejs/apm-agent-nodejs-mbp/PR-{number}&transactionType=unknown&latencyAggregationType=avg&traceId=95b38292497a6aa5dd231c3b7f8f86fa&transactionId=8df6f46581735f5b)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats



* Reason: `Aborted from #2`


* Start Time: 2021-08-09T20:41:11.918+0000


* Duration: 15 min 30 sec


* Commit: f8bf9bc30050ad4104d2843ab8aa2db5a9c2e2a2



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 2  |
| Skipped      | 0 |
| Total        | 2   |


#### Trends :test_tube:


![Image of Build Times](https://apm-ci.elastic.co/job/apm-agent-nodejs/job/apm-agent-nodejs-mbp/job/PR-2185/buildTimeGraph/png)


![Image of Tests](https://apm-ci.elastic.co/job/apm-agent-nodejs/job/apm-agent-nodejs-mbp/job/PR-2185/1/testReport/history/countGraph/png?start=0&amp;end=25)


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->









### Steps errors [![6](https://img.shields.io/badge/6%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-2185/detail/PR-2185/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Print Message`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2185/runs/1/steps/195/log/?start=0">here</a></li>
<li>Description: <code>[39;49m[INFO] Allocating a worker with the labels 'linux && immutable && extra/ac563d28-b219-47a5-adb9-e86130c3d126'.[0m</code></l1>
</ul>
  
  
  
  
##### `Print Message`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2185/runs/1/steps/197/log/?start=0">here</a></li>
<li>Description: <code>[39;49m[INFO] Allocating a worker with the labels 'linux && immutable && extra/dc62807a-9415-429e-a0fb-f5ba2bf3c9f3'.[0m</code></l1>
</ul>
  
  
  
  
##### `Print Message`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2185/runs/1/steps/199/log/?start=0">here</a></li>
<li>Description: <code>[39;49m[INFO] Allocating a worker with the labels 'linux && immutable && extra/0d897dca-47fa-4606-b138-930b3a85eb20'.[0m</code></l1>
</ul>
  
  
  
  
##### `Print Message`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2185/runs/1/steps/201/log/?start=0">here</a></li>
<li>Description: <code>[39;49m[INFO] Allocating a worker with the labels 'linux && immutable && extra/a6aafd28-8ce8-4ae1-90f0-0bcade6e4ad2'.[0m</code></l1>
</ul>
  
  
  
  
##### `Print Message`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2185/runs/1/steps/207/log/?start=0">here</a></li>
<li>Description: <code>[39;49m[INFO] Allocating a worker with the labels 'linux && immutable && extra/35913834-cd85-40fe-b11a-1063caa64e22'.[0m</code></l1>
</ul>
  
  
  
  
##### `Recursively delete the current directory from the workspace`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2185/runs/1/steps/354/log/?start=0">here</a></li>

</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2021-08-09T20:56:29.208Z]                 "EndpointID": "97cf27b7cbc883d99a48c910f00976c4b7d5db49f50bf4164eef32c03c6d0241",
[2021-08-09T20:56:29.208Z]                 "MacAddress": "02:42:ac:12:00:0a",
[2021-08-09T20:56:29.208Z]                 "IPv4Address": "172.18.0.10/16",
[2021-08-09T20:56:29.208Z]                 "IPv6Address": ""
[2021-08-09T20:56:29.208Z]             },
[2021-08-09T20:56:29.208Z]             "6628959355c73141b00fedec7f5b4ae1af0f6568a5766427133c6d5b27636e59": {
[2021-08-09T20:56:29.208Z]                 "Name": "docker_memcached_1",
[2021-08-09T20:56:29.208Z]                 "EndpointID": "f26b0916d96e3389dbe51b3dda49cf791373306ff3dbb2af25deb8e97467f2dc",
[2021-08-09T20:56:29.208Z]                 "MacAddress": "02:42:ac:12:00:07",
[2021-08-09T20:56:29.208Z]                 "IPv4Address": "172.18.0.7/16",
[2021-08-09T20:56:29.208Z]                 "IPv6Address": ""
[2021-08-09T20:56:29.208Z]             },
[2021-08-09T20:56:29.208Z]             "734c0b590a908ecfce5ffec5124a4f029b7f4a0546f38c78d19378c361a43a9d": {
[2021-08-09T20:56:29.208Z]                 "Name": "docker_redis_1",
[2021-08-09T20:56:29.208Z]                 "EndpointID": "98492d167faf4a92654e97ce20ca5044678ac0b6ce78775c487547e3607cb446",
[2021-08-09T20:56:29.208Z]                 "MacAddress": "02:42:ac:12:00:05",
[2021-08-09T20:56:29.208Z]                 "IPv4Address": "172.18.0.5/16",
[2021-08-09T20:56:29.208Z]                 "IPv6Address": ""
[2021-08-09T20:56:29.208Z]             },
[2021-08-09T20:56:29.208Z]             "7815933a96f54e133d963ebc1510ddc74d6f869fa9511522acc3cbd1928543d8": {
[2021-08-09T20:56:29.208Z]                 "Name": "docker_mysql_1",
[2021-08-09T20:56:29.208Z]                 "EndpointID": "240f0564bb571cac9c48ff31216f2d731a5c5d98b4d644a2a47429de13b08427",
[2021-08-09T20:56:29.208Z]                 "MacAddress": "02:42:ac:12:00:04",
[2021-08-09T20:56:29.208Z]                 "IPv4Address": "172.18.0.4/16",
[2021-08-09T20:56:29.208Z]                 "IPv6Address": ""
[2021-08-09T20:56:29.208Z]             },
[2021-08-09T20:56:29.208Z]             "92760f914aeea8af8d5b07e3390a9392694dd2f0d8fe89ae66672691dca6ff8d": {
[2021-08-09T20:56:29.208Z]                 "Name": "docker_localstack_1",
[2021-08-09T20:56:29.208Z]                 "EndpointID": "1e5185840eb82c528ad9fd15337d5923e96abbef4f5ba22be08fc1552eb6fe6f",
[2021-08-09T20:56:29.208Z]                 "MacAddress": "02:42:ac:12:00:02",
[2021-08-09T20:56:29.208Z]                 "IPv4Address": "172.18.0.2/16",
[2021-08-09T20:56:29.208Z]                 "IPv6Address": ""
[2021-08-09T20:56:29.208Z]             },
[2021-08-09T20:56:29.208Z]             "d8c0a743f955309d326d05cd7620487de10b8c985051c21eb7c52ffcf0d97a4a": {
[2021-08-09T20:56:29.208Z]                 "Name": "docker_postgres_1",
[2021-08-09T20:56:29.208Z]                 "EndpointID": "433a803e730ad9479c7e76cb148852da8d7d93499985a9348be1785356f61350",
[2021-08-09T20:56:29.208Z]                 "MacAddress": "02:42:ac:12:00:03",
[2021-08-09T20:56:29.208Z]                 "IPv4Address": "172.18.0.3/16",
[2021-08-09T20:56:29.208Z]                 "IPv6Address": ""
[2021-08-09T20:56:29.208Z]             },
[2021-08-09T20:56:29.208Z]             "dc778b31abe8d709c5b4562ac352ba913dcabc669f17836971b38377d064c7c7": {
[2021-08-09T20:56:29.208Z]                 "Name": "docker_mongodb_1",
[2021-08-09T20:56:29.208Z]                 "EndpointID": "5b4984a6562ccfb767235ec712d89ba18acc14aa42643314e93843481ef8f641",
[2021-08-09T20:56:29.208Z]                 "MacAddress": "02:42:ac:12:00:06",
[2021-08-09T20:56:29.208Z]                 "IPv4Address": "172.18.0.6/16",
[2021-08-09T20:56:29.208Z]                 "IPv6Address": ""
[2021-08-09T20:56:29.208Z]             }
[2021-08-09T20:56:29.208Z]         },
[2021-08-09T20:56:29.208Z]         "Options": {},
[2021-08-09T20:56:29.208Z]         "Labels": {
[2021-08-09T20:56:29.208Z]             "com.docker.compose.network": "default",
[2021-08-09T20:56:29.208Z]             "com.docker.compose.project": "docker",
[2021-08-09T20:56:29.208Z]             "com.docker.compose.version": "1.25.5"
[2021-08-09T20:56:29.208Z]         }
[2021-08-09T20:56:29.208Z]     }
[2021-08-09T20:56:29.208Z] ]
[2021-08-09T20:56:29.208Z] + docker network inspect -f '{{range .Containers}}{{ .Name }} {{end}}' docker_default
[2021-08-09T20:56:29.208Z] docker_elasticsearch_1 docker_cassandra_1 docker_mssql_1 docker_memcached_1 docker_redis_1 docker_mysql_1 docker_localstack_1 docker_postgres_1 docker_mongodb_1 
[2021-08-09T20:56:29.208Z] + docker network prune --force
[2021-08-09T20:56:29.603Z] + pwd
[2021-08-09T20:56:29.604Z] + id -u
[2021-08-09T20:56:29.604Z] + id -g
[2021-08-09T20:56:29.604Z] + docker run --rm -v /var/lib/jenkins/workspace/PR-2185-1-41a988f5-3302-4b98-b06d-22e3717777f1/src/github.com/elastic/apm-agent-nodejs:/usr/src/app -w /usr/src/app -u 1168:1169 node:12-alpine sh -c export HOME=/tmp ; mkdir ~/.npm-global; npm config set prefix ~/.npm-global ; npm install tap-xunit -g ; for i in "*-output.tap" ; do cat ${i} | /tmp/.npm-global/bin/tap-xunit --package="Node.js" > ${i%.*}-junit.xml ; done
[2021-08-09T20:56:29.604Z] Unable to find image 'node:12-alpine' locally
[2021-08-09T20:56:30.173Z] 12-alpine: Pulling from library/node
[2021-08-09T20:56:30.433Z] ddad3d7c1e96: Already exists
[2021-08-09T20:56:30.433Z] 2fe0e3281b4e: Already exists
[2021-08-09T20:56:30.433Z] b0460c564ff8: Already exists
[2021-08-09T20:56:30.433Z] 100581ca19ee: Already exists
[2021-08-09T20:56:30.696Z] Digest: sha256:78be4f61c7a0f00cc9da47e3ba2f1bacf9ba85410e8ad2e7de16e1fc4be1b962
[2021-08-09T20:56:30.696Z] Status: Downloaded newer image for node:12-alpine
[2021-08-09T20:56:35.977Z] /tmp/.npm-global/bin/tap-xunit -> /tmp/.npm-global/lib/node_modules/tap-xunit/bin/tap-xunit
[2021-08-09T20:56:35.977Z] /tmp/.npm-global/bin/txunit -> /tmp/.npm-global/lib/node_modules/tap-xunit/bin/tap-xunit
[2021-08-09T20:56:35.977Z] + tap-xunit@2.4.1
[2021-08-09T20:56:35.977Z] added 21 packages from 21 contributors in 2.218s
[2021-08-09T20:56:36.260Z] Recording test results
[2021-08-09T20:56:37.021Z] [Checks API] No suitable checks publisher found.
[2021-08-09T20:56:37.363Z] Terminated
[2021-08-09T20:56:37.861Z] [INFO] getVaultSecret: Getting secrets
[2021-08-09T20:56:37.914Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-08-09T20:56:39.680Z] Stage "TAV Test" skipped due to earlier failure(s)
[2021-08-09T20:56:39.778Z] Stage "Edge Test" skipped due to earlier failure(s)
[2021-08-09T20:56:39.952Z] Stage "Nightly Test" skipped due to earlier failure(s)
[2021-08-09T20:56:39.965Z] Stage "Nightly Test - No async hooks" skipped due to earlier failure(s)
[2021-08-09T20:56:39.979Z] Stage "RC Test" skipped due to earlier failure(s)
[2021-08-09T20:56:39.991Z] Stage "RC Test - No async hooks" skipped due to earlier failure(s)
[2021-08-09T20:56:40.146Z] Failed in branch Nightly Test
[2021-08-09T20:56:40.156Z] Failed in branch Nightly Test - No async hooks
[2021-08-09T20:56:40.166Z] Failed in branch RC Test
[2021-08-09T20:56:40.175Z] Failed in branch RC Test - No async hooks
[2021-08-09T20:56:40.412Z] Stage "Integration Tests" skipped due to earlier failure(s)
[2021-08-09T20:56:40.509Z] Stage "Release" skipped due to earlier failure(s)
[2021-08-09T20:56:40.549Z] Stage "Release" skipped due to earlier failure(s)
[2021-08-09T20:56:40.701Z] Stage "Benchmarks" skipped due to earlier failure(s)
[2021-08-09T20:56:41.006Z] Running on Jenkins in /var/lib/jenkins/workspace/ejs_apm-agent-nodejs-mbp_PR-2185
[2021-08-09T20:56:41.104Z] [INFO] getVaultSecret: Getting secrets
[2021-08-09T20:56:41.160Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-08-09T20:56:42.071Z] + chmod 755 generate-build-data.sh
[2021-08-09T20:56:42.071Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2185/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2185/runs/1 ABORTED 929882
[2021-08-09T20:56:42.071Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2185/runs/1/steps/?limit=10000 -o steps-info.json

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

trentm commented on lib/lambda.js:
```diff
@@ -86,11 +87,14 @@ module.exports = function elasticApmAwsLambda (agent) {
              * elastic-apm-traceparent if available.
              */
             parentId = value
```
1. The handling will have to change to avoid that `break` preferring 'elastic-apm-traceparent' to 'traceparent'. IIUC, with the current logic it will miss tracestate if it comes after 'elastic-apm-traceparent' in `payload.headers`.
2. Also, per #2079 the logic should change (could be a separate PR) to prefer 'traceparent' to 'elastic-apm-traceparent'.

astorm replied:
@trentm Ugh -- that's a weird construct.  Thanks for catching it. Any strong preference between 

- more code and fewer loop cycles (matching most closely with current intent/pattern)

```
    if (payload.headers !== undefined) {
      for (const [key, value] of Object.entries(payload.headers)) {
        const lowerCaseKey = key.toLowerCase()
        if (lowerCaseKey === 'traceparent') {
          parentId = value
        } else if (!parentId && lowerCaseKey === 'elastic-apm-traceparent') {
          parentId = value
        } else if(lowerCaseKey === 'tracestate') {
          tracestate = value
        }
        if(parentId && tracestate) {
          break
        }
      }
    }
```

- or less code but slightly more work normalizing all the headers

```
      if (payload.headers !== undefined) {
        const normalizedHeaders = {}
        for (const [key, value] of Object.entries(payload.headers)) {
          const lowerCaseKey = key.toLowerCase()
          normalizedHeaders[lowerCaseKey] = value
        }
        parentId = normalizedHeaders.traceparent ? normalizedHeaders.traceparent : normalizedHeaders['elastic-apm-traceparent']
        tracestate = normalizedHeaders.tracestate
      }
```

trentm replied:
The former will prefer 'elastic-apm-traceparent' over 'traceparent' if it and 'tracestate' come first, so doesn't work as written.

No strong preference between those two until perf profiles show that the section is a hot spot.

Slightly related: from old https://elastic.slack.com/archives/C5M02D1GE/p1619802623086700?thread_ts=1619742610.083400&cid=C5M02D1GE `for ([key, value] of Object.entries(obj))` is dog slow. Perhaps premature optimization, but I'd generally bias against using it.

astorm commented:
I've replaced the `Object.entries()` loop with an `Object.keys()` loop. Always up for quashing these then we see them and remember.

I've opted for the clearer (to me) code of normalizing _all_ the headers rather than the micro-optimization of finishing once we have the specific headers we're interested in. 

I've added test cases with both `elastic-apm-traceparent` by itself and `elastic-apm-traceparent` both before and after the `traceparent` header to ensure that we prefer `traceparent` when it's there, but still pickup `elastic-apm-traceparent` when it's by itself.

trentm reviewed (approved):
Thanks.
